marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -1,185 +1,215 @@
+import RewindableIndex.Storable qualified as Storable
 
-$(makeLenses ''Utxo)
+type UtxoIndexer = Storable.State UtxoHandle
 
-instance Eq Utxo where
-    u1 == u2 = (_utxoTxId u1) == (_utxoTxId u2)
+data UtxoHandle = UtxoHandle
+  { hdlConnection :: SQL.Connection
+  , hdlDpeth      :: Int
+  }
 
-instance Ord Utxo where
-    compare u1 u2 = compare (_utxoTxId u1) (_utxoTxId u2)
+newtype instance StorableQuery UtxoHandle =
+  UtxoAddress C.AddressAny deriving (Show, Eq)
 
-data UtxoEvent          = UtxoEvent
-    { _utxoEventUtxos   :: [Utxo]
-    , _utxoEventInputs  :: !(Set C.TxIn)
-    , _utxoEventSlotNo  :: !C.SlotNo
-    , _utxoEventBlockNo :: !C.BlockNo
-    } deriving (Show, Eq)
+type QueryableAddresses = NonEmpty (StorableQuery UtxoHandle)
 
-$(makeLenses ''UtxoEvent)
+type instance StorableMonad UtxoHandle = IO
 
-data UtxoRow            = UtxoRow
-    { _utxoRowUtxo    :: Utxo
-    , _utxoRowSlotNo  :: !C.SlotNo
-    , _utxoRowBlockNo :: !C.BlockNo
-    } deriving (Show, Eq, Ord, Generic)
+type instance StorablePoint UtxoHandle = C.ChainPoint
 
-$(makeLenses ''UtxoRow)
+newtype Depth = Depth Int
 
-type Result = Maybe [UtxoRow]
+data Utxo = Utxo
+  { _address          :: StorableQuery UtxoHandle
+  , _txId             :: !C.TxId
+  , _txIx             :: !C.TxIx
+  , _datum            :: Maybe C.ScriptData
+  , _datumHash        :: Maybe (C.Hash C.ScriptData)
+  , _value            :: C.Value
+  , _inlineScript     :: Maybe ByteString
+  , _inlineScriptHash :: Maybe C.ScriptHash
+  } deriving (Show, Generic)
 
-type Notification = ()
+$(makeLenses ''Utxo)
 
-type UtxoIndex
-    = SqliteIndex
-      UtxoEvent
-      Notification
-      C.AddressAny
-      Result
+instance Aeson.ToJSON C.ScriptData where
+  toJSON = C.scriptDataToJson C.ScriptDataJsonDetailedSchema
 
-newtype Depth = Depth Int
+instance Eq Utxo where
+  left == right
+    =  _txId left == _txId right
+    && _txIx left == _txIx right
+
+instance Ord Utxo where
+  left <= right = _txId left <= _txId right
+
+data UtxoRow = UtxoRow
+  { _urUtxo      :: Utxo
+  , _urBlockNo   :: C.BlockNo
+  , _urSlotNo    :: C.SlotNo
+  , _urBlockHash:: C.Hash C.BlockHeader
+  } deriving (Show, Eq, Ord, Generic)
+
+$(makeLenses ''UtxoRow)
+
+newtype instance  StorableResult UtxoHandle = UtxoResult [UtxoRow] deriving Show
+
+data instance  StorableEvent UtxoHandle = UtxoEvent
+  { ueUtxos       :: !(Set Utxo)
+  , ueInputs      :: Set C.TxIn
+  , ueBlockNo     :: C.BlockNo
+  , ueChainPoint  :: C.ChainPoint
+  } deriving (Show, Generic)
+
+eventIsBefore :: StorableEvent UtxoHandle  -> C.ChainPoint -> Bool
+eventIsBefore (UtxoEvent _ _ _ (C.ChainPoint _ slot)) (C.ChainPoint _ slot') =  slot < slot'
+eventIsBefore _ _                                                            = False
+
+instance Semigroup (StorableEvent UtxoHandle) where
+  (UtxoEvent u i b cp) <> (UtxoEvent u' i' _ cp') =
+    if cp /= cp' then
+      error "can not combine UtxoEvents from different ChainPoints"
```

Comment:
An `error` in the instance? Mmm... I don't like that :D 
Is that common practice?

Reply:
no, I changed it.